src/understand-tidb/implementation-of-vectorized-execution.md

```diff
@@ -1 +1,125 @@
+- A `0` is appended to the `nullBitmap`.
+- If it's a fixed-length column, a placeholder data of the same size as `elemBuf` is appended to the `data`.
+- If it's a variable-length column, no data is appended to the `data`, but the starting point of the next element is recorded in the `offsets`.
+
+**Appending a NULL Value to a Column**:
+- To append a NULL value, the [AppendNull](https://github.com/pingcap/tidb/blob/ecaa1c518cc9367844ebb5206f2e970461c8bf28/pkg/util/chunk/column.go#L229-L237) function is used.
+- A `0` is appended to the `nullBitmap`.
+- If it's a fixed-length column, a placeholder data of the same size as `elemBuf` is appended to the `data`.
+- If it's a variable-length column, no data is appended to the `data`, but the starting point of the next element is recorded in the `offsets`.
+
+**Reading a Value from a Column**:
+- Values in a `Column` can be read using functions like [GetInt64(rowIdx)](https://github.com/pingcap/tidb/blob/ecaa1c518cc9367844ebb5206f2e970461c8bf28/pkg/util/chunk/column.go#L551-L553) and [GetString(rowIdx)](https://github.com/pingcap/tidb/blob/ecaa1c518cc9367844ebb5206f2e970461c8bf28/pkg/util/chunk/column.go#L576-L578). The reading principle can be deduced from the earlier-described appending mechanism. Here, we retrieve the specified element in the `Column` based on the rowID. The details of reading from a `Column` are consistent with the principles discussed for appending.
+
+**Reading a Row from a Chunk**:
+- Within a `Chunk`, the concept of a [Row](https://github.com/pingcap/tidb/blob/ecaa1c518cc9367844ebb5206f2e970461c8bf28/pkg/util/chunk/row.go#L25-L28) is logical. The data for a row is stored across different `Columns` in the `Chunk`. The data for the same row in various columns is not necessarily stored consecutively in memory. When obtaining a `Row` object, there is no need to perform data copying, as the data for the same row is already stored in the corresponding `Columns`.
+- The concept of a `Row` is useful because, during the operation of operators, data is often accessed and processed on a per-row basis. For example, operations like aggregation, sorting, and similar tasks work with data at the row level.
+- You can retrieve a row from a `Chunk` using the [GetRow(rowIdx)](https://github.com/pingcap/tidb/blob/ecaa1c518cc9367844ebb5206f2e970461c8bf28/pkg/util/chunk/chunk.go#L356-L368) function. Once you have a `Row` object, you can further access the data in specific columns within that row using functions like [Row::GetInt64(colIdx)](https://github.com/pingcap/tidb/blob/ecaa1c518cc9367844ebb5206f2e970461c8bf28/pkg/util/chunk/row.go#L51-L53), which allows you to retrieve the data corresponding to the specified column index for that row.
+
+### Example: How to execute `select a+1, b-10 from t where a%2=0` using vectorized execution
```

Comment:
I think we should add two sections here: 1) how expression is evaluated in a vectorized fashion; 2) how certain operator exploits vectorized processing.

For expression, maybe we can pick up some key point from https://www.pingcap.com/blog/10x-performance-improvement-for-expression-evaluation-made-possible-by-vectorized-execution/ (and reference this article for more details).

For operators, maybe we can give an example of join probe, showing how vectorized processing differs with the row at a time, specifically: the steps of probing the join key in batch, gathering the matching rows from each output column to compose the output batch.